<a href="https://colab.research.google.com/github/samreenkauser/Rebounce-Applied-AI-and-Analytics-Codes/blob/main/Python_301_ipynb.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
import numpy as np

url = "https://huggingface.co/datasets/aarav912/online-retail/resolve/main/online_retail.csv"
df = pd.read_csv(url)

In [2]:
df.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom


In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


In [4]:
df.isnull().sum()

,0
InvoiceNo,0
StockCode,0
Description,1454
Quantity,0
InvoiceDate,0
UnitPrice,0
CustomerID,135080
Country,0


In [5]:
df = df.dropna(subset=["CustomerID"])

In [6]:
df["InvoiceDate"] = pd.to_datetime(df["InvoiceDate"])

In [7]:
df["TotalAmount"] = df["Quantity"] * df["UnitPrice"]

In [8]:
df["Sales"] = df["Quantity"] * df["UnitPrice"]

In [9]:
df = df.drop_duplicates()

In [10]:
df = df[(df["Quantity"] > 0) & (df["UnitPrice"] > 0)]

In [11]:
sales_above_500 = df[df["Sales"] > 500]
print("Number of transactions above 500:", len(sales_above_500))

Number of transactions above 500: 913


In [12]:
top_10_transactions = df.sort_values("Sales", ascending=False).head(10)
print(top_10_transactions)

       InvoiceNo StockCode                          Description  Quantity  \
540421    581483     23843          PAPER CRAFT , LITTLE BIRDIE     80995   
61619     541431     23166       MEDIUM CERAMIC TOP STORAGE JAR     74215   
222680    556444     22502       PICNIC BASKET WICKER 60 PIECES        60   
173382    551697      POST                              POSTAGE         1   
348325    567423     23243  SET OF TEA COFFEE SUGAR TINS PANTRY      1412   
160546    550461     21108   FAIRY CAKE FLANNEL ASSORTED COLOUR      3114   
52711     540815     21108   FAIRY CAKE FLANNEL ASSORTED COLOUR      3114   
421601    573003     23084                   RABBIT NIGHT LIGHT      2400   
52709     540815    85123A   WHITE HANGING HEART T-LIGHT HOLDER      1930   
160542    550461    85123A   WHITE HANGING HEART T-LIGHT HOLDER      1930   

               InvoiceDate  UnitPrice  CustomerID         Country  \
540421 2011-12-09 09:15:00       2.08     16446.0  United Kingdom   
61619  2011-01

In [13]:
germany = df[df["Country"] == "Germany"]
print("Number of transactions from Germany:", len(germany))

Number of transactions from Germany: 9025


In [14]:
print("Mean:", np.mean(df["Sales"]))
print("Median:",
      np.median(df["Sales"]))
print("Minimum:",
      np.min(df["Sales"]))
print("Maximum:",
      np.max(df["Sales"]))
print("Standard Deviation:",
      np.std(df["Sales"]))

Mean: 22.6314997351614
Median: 12.45
Minimum: 0.001
Maximum: 168469.6
Standard Deviation: 311.09882822360305


### Numpy Analysis
  The mean sales value is higher than the median sales value. This indicates that the sales distribution is positively skewed, because a small number of very large transactions iincrease the mean. The maximum sales value is much higher than the median, showing the presence of unusually large transactions.

In [15]:
large_quantity = df[df["Quantity"] > 1000]
print(large_quantity)

       InvoiceNo StockCode                        Description  Quantity  \
4850      536809     84950     ASSORTED COLOUR T-LIGHT HOLDER      1824   
4945      536830     84077  WORLD WAR 2 GLIDERS ASSTD DESIGNS      2880   
4946      536830     21915             RED  HARMONICA IN BOX       1400   
6365      536890    17084R              ASSORTED INCENSE PACK      1440   
16435     537659     22189            CREAM HEART CARD HOLDER      1008   
...          ...       ...                                ...       ...   
533812    581115     22413    METAL SIGN TAKE IT OR LEAVE IT       1404   
534952    581175     23084                 RABBIT NIGHT LIGHT      1440   
540070    581458     22197                     POPCORN HOLDER      1500   
540071    581459     22197                     POPCORN HOLDER      1200   
540421    581483     23843        PAPER CRAFT , LITTLE BIRDIE     80995   

               InvoiceDate  UnitPrice  CustomerID         Country  \
4850   2010-12-02 16:48:00    

In [16]:
summary = df.groupby("Country").agg(Number_of_Transactions=("Sales", "count"),
                                    Total_Sales=("Sales", "sum"),
                                    Average_Sales=("Sales", "mean")).reset_index()
print(summary)

                 Country  Number_of_Transactions  Total_Sales  Average_Sales
0              Australia                    1181   138453.810     117.234386
1                Austria                     398    10198.680      25.624824
2                Bahrain                      17      548.400      32.258824
3                Belgium                    2031    41196.340      20.283772
4                 Brazil                      32     1143.600      35.737500
5                 Canada                     151     3666.380      24.280662
6        Channel Islands                     747    20440.540      27.363507
7                 Cyprus                     603    13502.850      22.392786
8         Czech Republic                      25      826.740      33.069600
9                Denmark                     380    18955.340      49.882474
10                  EIRE                    7226   265262.460      36.709446
11    European Community                      60     1300.250      21.670833

In [17]:
top_5_countries = summary.sort_values("Total_Sales", ascending=False).head(5)
print(top_5_countries)

           Country  Number_of_Transactions  Total_Sales  Average_Sales
35  United Kingdom                  349203  7285024.644      20.861862
23     Netherlands                    2359   285446.340     121.003111
10            EIRE                    7226   265262.460      36.709446
14         Germany                    9025   228678.400      25.338327
13          France                    8326   208934.310      25.094200


In [18]:
highest_average = summary.sort_values("Average_Sales", ascending=False).head(1)
print(highest_average)

        Country  Number_of_Transactions  Total_Sales  Average_Sales
23  Netherlands                    2359    285446.34     121.003111


In [19]:
top_country = summary.sort_values("Total_Sales", ascending=False).iloc[0]

print("Top-selling country:", top_country["Country"])
print("Number of transactions:", top_country["Number_of_Transactions"])

Top-selling country: United Kingdom
Number of transactions: 349203


In [20]:
region_lookup = {
    "United Kingdom": "Europe",
    "France": "Europe",
    "Germany": "Europe",
    "Spain": "Europe",
    "Netherlands": "Europe",
    "Belgium": "Europe",
    "Switzerland": "Europe",
    "Portugal": "Europe",
    "Australia": "Oceania",
    "Japan": "Asia",
    "USA": "Americas"}

regions = pd.DataFrame(list(region_lookup.items()), columns=["Country", "Region"])

print(regions)

           Country    Region
0   United Kingdom    Europe
1           France    Europe
2          Germany    Europe
3            Spain    Europe
4      Netherlands    Europe
5          Belgium    Europe
6      Switzerland    Europe
7         Portugal    Europe
8        Australia   Oceania
9            Japan      Asia
10             USA  Americas


In [21]:
summary = pd.merge(summary, regions, on="Country", how="left")

summary["Region"] = summary["Region"].fillna("other")

print(summary)

                 Country  Number_of_Transactions  Total_Sales  Average_Sales  \
0              Australia                    1181   138453.810     117.234386   
1                Austria                     398    10198.680      25.624824   
2                Bahrain                      17      548.400      32.258824   
3                Belgium                    2031    41196.340      20.283772   
4                 Brazil                      32     1143.600      35.737500   
5                 Canada                     151     3666.380      24.280662   
6        Channel Islands                     747    20440.540      27.363507   
7                 Cyprus                     603    13502.850      22.392786   
8         Czech Republic                      25      826.740      33.069600   
9                Denmark                     380    18955.340      49.882474   
10                  EIRE                    7226   265262.460      36.709446   
11    European Community                

In [22]:
region_summary = summary.groupby("Region").agg(
    Number_of_Transactions=("Number_of_Transactions", "sum"),
    Total_Sales=("Total_Sales", "sum"),
    Average_Sales=("Average_Sales", "mean")
).reset_index()

print(region_summary)

     Region  Number_of_Transactions  Total_Sales  Average_Sales
0  Americas                     179     3580.390      20.002179
1      Asia                     321    37416.370     116.561900
2    Europe                  376717  8200658.384      36.380372
3   Oceania                    1181   138453.810     117.234386
4     other                   14294   507099.940      33.472775


In [23]:
region_sales = summary.groupby("Region")["Total_Sales"].sum().sort_values(ascending=False)

print(region_sales)

Region
Europe      8200658.384
other        507099.940
Oceania      138453.810
Asia          37416.370
Americas       3580.390
Name: Total_Sales, dtype: float64


In [24]:
print("Best-performing region:", region_sales.idxmax())
print("Total sales:", region_sales.max())

Best-performing region: Europe
Total sales: 8200658.384000001


1. The UK generated the highest total sales, approx 7,285,024.64.
2. The Netherlands had the highest average transaction value, approx 121.0 per transaction.
3. Europe performed best, generating approx 8,200,658.38 in total sales.
4. The UK generated the highest total sales,becoz it had a very large number of transactions,while the Netherlanda had a much higher average transaction value.
5. Missing CustomerID values were removed. Duplicate records & transactions with non-positive Quantity were also removed.


Final Insight -

  The UK generated the highest total sales, while Europe was the best-performing region overall. The Netherlands had the highest average transaction value despite having much lower total sales than the UK. The mean sales valaue was higher than the median, indicatiing that a small number of very large transactions affected the overall distribution.

## Data Analysis Summary

   - The dataset contains online retail transaction data.
   - Missing Customer ID values were removed.
   - InvoiceDate was converted to datetime format.
   - Duplicate records were removed.
   - Transactions with Quantity and UnitPrice greater than 0 were retained.
   - Total sales were calculated as Quantity * UnitPrice.
   - Europe is the best-performing region based on totala sales.